Upstream - patch dynamic provisioning

markturansky opened the pull request:
This PR includes, in order of commits:

1.  new PersistentVolumeController
2.  updates to Origin to support new controller
3. Cinder provisioner
4. GCE provisioner
5. EBS provisioner

@jsafrane

liggitt commented:
Normal upstream commits with matching pull requests don't need to specify `drop`… that's really just for ones without a match that should specify drop or carry

jsafrane commented:
It does not compile, Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/ misses some files.

markturansky commented:
@jsafrane git settings prevented new files from being added.  they are added now and build correctly.

markturansky commented:
@liggitt These 4 upstream PRs are not yet merged.  Should they be marked <carry> until they are merged upstream and can come in during a normal rebase?

Local testing is successful for me.  Jan's reported 2 of 3 plugins validated so far.

We have https://github.com/openshift/origin/issues/3286 for a specific service account for pv.    I'm unfamiliar with this code, so if you could help with the technical design in that issue, I will add the service account in a separate PR.

liggitt commented:
> Should they be marked <carry> until they are merged upstream and can come in during a normal rebase?

No, just reference the upstream PR. During a rebase, we inspect the referenced PRs. If they have merged, we drop the commit. If they have been closed without merging, we either drop or change the title to `<carry>`, whichever is appropriate.

markturansky commented:
Do i need to edit the commits to remove <drop> or are they ok?

markturansky commented:
er, to remove < drop >

liggitt commented:
remove `<drop>`

markturansky commented:
will do

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/types.go:
```diff
@@ -26,10 +26,24 @@ import (
+	provisionedForKey = "volume.experimental.kubernetes.io/provisioned-for"
+	// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
+	// with namespace of a persistent volume claim used to create this volume.
+	cloudVolumeCreatedForNamespaceTag = "kubernetes.io/created-for/PVC/Namespace"
```
lowercase?

markturansky replied:
I'd prefer to have this kind of feedback on the upstream PR than here, unless all lowercase is required in OpenShift for a reason I'm unaware of.

liggitt replied:
not openshift specific, I'll add it upstream

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/persistentvolume/types.go:
```diff
@@ -26,10 +26,24 @@ import (
+	cloudVolumeCreatedForNamespaceTag = "kubernetes.io/created-for/PVC/Namespace"
+	// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
+	// with name of a persistent volume claim used to create this volume.
+	cloudVolumeCreatedForNameTag = "kubernetes.io/created-for/PVC/Name"
```
lowercase?

markturansky commented:
@liggitt `drop` removed from commits.

Between amending an earlier commit to add missing files and then editing each of the commit messages in order, I can say I learned some new git-fu with this PR.

markturansky commented:
Jan tested all the provisioning plugins (gce/ebs/cinder) and I verified nfs recycling via the wordpress example.

Improving annotation values notwithstanding, I think this PR is good-to-go.

markturansky commented on pkg/cmd/server/kubernetes/master.go:
```diff
@@ -58,41 +63,65 @@ func (c *MasterConfig) RunNamespaceController() {
+	// https://github.com/openshift/origin/issues/3286
+//	if err := attemptToLoadRecycler(volumeConfig.PersistentVolumeRecyclerPodTemplateFilePathHostPath, &hostPathConfig); err != nil {
+//		glog.Fatalf("Could not create hostpath recycler pod from file %s: %+v", volumeConfig.PersistentVolumeRecyclerPodTemplateFilePathHostPath, err)
+//	}
```
@liggitt The configurable recycler pods are removed pending the introduction of a service account.

markturansky commented:
@liggitt 2 changes

1.  I had to add missing files (see Jan's 1st comment) but they got added to the wrong commit (2nd instead of 1st).  I just got these files into the 1st commit where they belong.

2.  I removed all the code that loads the recycler pod, pending introduction of a service account.  I had it commented out, but removed it instead and also removed their now-unused helper method.  Code is clean.

Would you please review this again today and help me get this merged?

markturansky commented:
@childsb

liggitt commented:
@markturansky how close are the upstream PRs to LGTM? I'd rather not take a change as big as this when there's likelihood of significant drift

liggitt commented on pkg/cmd/server/kubernetes/master.go:
```diff
@@ -58,39 +62,39 @@ func (c *MasterConfig) RunNamespaceController() {
-	binder.Run()
-}
-
 func (c *MasterConfig) RunPersistentVolumeClaimRecycler(recyclerImageName string) {
```
change name to `RunPersistentVolumeController`

markturansky replied:
function name changed.

liggitt commented on pkg/cmd/server/kubernetes/master.go:
```diff
@@ -58,39 +62,39 @@ func (c *MasterConfig) RunNamespaceController() {
+	allPlugins = append(allPlugins, gce_pd.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, cinder.ProbeVolumePlugins()...)
+
+	qosClasses := c.ControllerManager.VolumeConfigFlags.ProvisionerQualityOfServiceClasses
```
`volumeConfig...`

markturansky replied:
changed to `storageClasses`, which is more accurate.  upstream prefers not to use the word "tiers".

These things are a subset of volumeConfig, though, so naming it volumeConfig would not be correct in this case.

smarterclayton replied:
This is too big to pull in for 1.1 without being merged upstream, with an
unknown amount of change left.  It'll have to be post 1.1.

On Thu, Oct 22, 2015 at 12:58 PM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/cmd/server/kubernetes/master.go
> <https://github.com/openshift/origin/pull/5217#discussion_r42775017>:
>
> >  		RecyclerPodTemplate:      defaultScrubPod,
> >  	}
> >
> >  	allPlugins := []volume.VolumePlugin{}
> >  	allPlugins = append(allPlugins, host_path.ProbeVolumePlugins(hostPathConfig)...)
> >  	allPlugins = append(allPlugins, nfs.ProbeVolumePlugins(nfsConfig)...)
> > -
> > -	recycler, err := volumeclaimbinder.NewPersistentVolumeRecycler(c.KubeClient, c.ControllerManager.PVClaimBinderSyncPeriod, allPlugins)
> > +	allPlugins = append(allPlugins, aws_ebs.ProbeVolumePlugins()...)
> > +	allPlugins = append(allPlugins, gce_pd.ProbeVolumePlugins()...)
> > +	allPlugins = append(allPlugins, cinder.ProbeVolumePlugins()...)
> > +
> > +	qosClasses := c.ControllerManager.VolumeConfigFlags.ProvisionerQualityOfServiceClasses
>
> changed to storageClasses, which is more accurate. upstream prefers not
> to use the word "tiers".
>
> These things are a subset of volumeConfig, though, so naming it
> volumeConfig would not be correct in this case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5217/files#r42775017>.
>

markturansky replied:
We've tested through all of it and it also fixes a race bug.

I don't expect a lot of upstream change, but I also cannot accurately guess how much they will want.  It might be 0.

smarterclayton replied:
Given the risk involved in other changes, I don't want to land this until
it's had bake time.  I don't want that bake time to be in Origin.

On Thu, Oct 22, 2015 at 2:25 PM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/cmd/server/kubernetes/master.go
> <https://github.com/openshift/origin/pull/5217#discussion_r42786220>:
>
> >  		RecyclerPodTemplate:      defaultScrubPod,
> >  	}
> >
> >  	allPlugins := []volume.VolumePlugin{}
> >  	allPlugins = append(allPlugins, host_path.ProbeVolumePlugins(hostPathConfig)...)
> >  	allPlugins = append(allPlugins, nfs.ProbeVolumePlugins(nfsConfig)...)
> > -
> > -	recycler, err := volumeclaimbinder.NewPersistentVolumeRecycler(c.KubeClient, c.ControllerManager.PVClaimBinderSyncPeriod, allPlugins)
> > +	allPlugins = append(allPlugins, aws_ebs.ProbeVolumePlugins()...)
> > +	allPlugins = append(allPlugins, gce_pd.ProbeVolumePlugins()...)
> > +	allPlugins = append(allPlugins, cinder.ProbeVolumePlugins()...)
> > +
> > +	qosClasses := c.ControllerManager.VolumeConfigFlags.ProvisionerQualityOfServiceClasses
>
> We've tested through all of it and it also fixes a race bug.
>
> I don't expect a lot of upstream change, but I also cannot accurately
> guess how much they will want. It might be 0.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5217/files#r42786220>.
>

liggitt commented:
a couple nits, origin code looks fine... I'd like someone involved in the upstream pulls to weigh in on pulling in the upstream commits in their current state

markturansky commented:
@liggitt thanks for the feedback. I'll have it online shortly.

As for upstream, the long and short of it is that the pair of existing controllers get complicated with the addition of a 3rd and I had a race condition between them to resolve.  This new single controller combines all 3, is simpler to follow, and isn't racy. Feedback from RH is in the new controller (tested by RH, too), but upstream has not yet weighed in. This new effort is less code than before, better documented, and works well.  I don't expect much upstream pushback.  

I think the window for changes has closed, though, if we want this in 3.1. I have some PTO coming up and devcut was yesterday, I think.

@smarterclayton thoughts?

liggitt commented:
I'm ok taking it before merge upstream, just asking how much more change was expected

markturansky commented:
Unknown amount of change.

markturansky commented:
Will create a new PR with the upstream version of provisioning when available.
